### PR TITLE
Add model arcinstitute state

### PR DESCRIPTION
**Pull request by drbh**
This PR adds the arc state model

# Run embedding model via transformers

```bash
git clone https://github.com/huggingface/transformers
git checkout add-model-arcinstitute-state
uv run sanity.py
```

sanity.py
```python
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "torch",
#     "transformers"
# ]
#
# [tool.uv.sources]
# transformers = { path = ".", editable = true }
# ///
import torch
from transformers import StateEmbeddingModel


model_name = "arcinstitute/SE-600M"
model = StateEmbeddingModel.from_pretrained(model_name)

torch.manual_seed(0)
input_ids = torch.randn((1, 1, 5120), dtype=torch.float32)
mask = torch.ones((1, 1, 5120), dtype=torch.bool)
mask[:, :, 2560:] = False # simulate half masking
print("Input sum:\t", input_ids.sum())
print("Mask sum:\t", mask.sum())

outputs = model(input_ids, mask)
print("Output sum:\t", outputs["gene_output"].sum())
```

outputs

```txt
Input sum:	 tensor(-38.6611)
Mask sum:	 tensor(2560)
Output sum:	 tensor(-19.6819, grad_fn=<SumBackward0>)
```

# Compare to reference

```
git clone https://github.com/ArcInstitute/state.git
cd state
curl -OL https://huggingface.co/arcinstitute/SE-600M/resolve/main/se600m_epoch16.ckpt
```

next, apply this small patch so we can run the model file directly with a fixed input to compare with the impl above

<details>
  <summary>file `compare.patch`</summary>
  
  ```patch
  diff --git a/src/state/emb/nn/model.py b/src/state/emb/nn/model.py
  index dbbefb3..42167a1 100644
  --- a/src/state/emb/nn/model.py
  +++ b/src/state/emb/nn/model.py
  @@ -23,20 +23,20 @@ from torch.nn import TransformerEncoder, TransformerEncoderLayer, BCEWithLogitsL
   from tqdm.auto import tqdm
   from torch.optim.lr_scheduler import ChainedScheduler, LinearLR, CosineAnnealingLR, ReduceLROnPlateau
   
  -from ..data import create_dataloader
  -from ..utils import (
  +from state.emb.data import create_dataloader
  +from state.emb.utils import (
       compute_gene_overlap_cross_pert,
       get_embedding_cfg,
       get_dataset_cfg,
       compute_pearson_delta,
       compute_perturbation_ranking_score,
   )
  -from ..eval.emb import cluster_embedding
  -from .loss import WassersteinLoss, KLDivergenceLoss, MMDLoss, TabularLoss
  +from state.emb.eval.emb import cluster_embedding
  +from loss import WassersteinLoss, KLDivergenceLoss, MMDLoss, TabularLoss
   
   
  -from .flash_transformer import FlashTransformerEncoderLayer
  -from .flash_transformer import FlashTransformerEncoder
  +from flash_transformer import FlashTransformerEncoderLayer
  +from flash_transformer import FlashTransformerEncoder
   
   
   class SkipBlock(nn.Module):
  @@ -196,7 +196,8 @@ class StateEmbeddingModel(L.LightningModule):
               self.dataset_embedder = nn.Linear(output_dim, 10)
   
               # Assume self.cfg.model.num_datasets is set to the number of unique datasets.
  -            num_dataset = get_dataset_cfg(self.cfg).num_datasets
  +            # num_dataset = get_dataset_cfg(self.cfg).num_datasets
  +            num_dataset = 14420 
               self.dataset_encoder = nn.Sequential(
                   nn.Linear(output_dim, d_model),
                   nn.SiLU(),
  @@ -686,3 +687,18 @@ class StateEmbeddingModel(L.LightningModule):
               "optimizer": optimizer,
               "lr_scheduler": {"scheduler": scheduler, "monitor": "train_loss", "interval": "step", "frequency": 1},
           }
  +
  +if __name__ == "__main__":
  +    checkpoint = "/Users/drbh/Projects/state/se600m_epoch16.ckpt"
  +    model = StateEmbeddingModel.load_from_checkpoint(checkpoint, dropout=0.0, strict=False)
  +
  +    torch.manual_seed(0)
  +
  +    input_ids = torch.randn((1, 1, 5120), dtype=torch.float32)
  +    mask = torch.ones((1, 1, 5120), dtype=torch.bool)
  +    mask[:, :, 2560:] = False
  +    print("Input sum:\t", input_ids.sum())
  +    print("Mask sum:\t", mask.sum())
  +
  +    output, embedding, dataset_emb = model(input_ids, mask)
  +    print("Output shape:\t", output.sum())
  
  ```
</details>

can be applied like

```bash
# save above as compare.patch
git apply compare.patch
```

run the model

```bash
.venv/bin/python src/state/emb/nn/model.py
```

output

```txt
!!! Using Flash Attention !!!
Input sum: tensor(-38.6611)
Mask sum: tensor(2560)
Output shape: tensor(-19.6819, grad_fn=<SumBackward0>)
```

### CODE REVIEW

**Comment by FL33TW00D** [comment]
Paper for reference! https://www.biorxiv.org/content/10.1101/2025.06.26.661135v2

**Comment by FL33TW00D** [comment]
@abhinadduri for ref

**Comment by ArthurZucker** [comment]
cc @Cyrilvallez camn you have a look?!

**Comment by Cyrilvallez** [comment]
Hey! I'm a bit confused with the PR right now! What model are we adding? If it's `arcinstitute`, we only want modeling file refering to it! No general names such as `StateEmbedding`!
But I see that you incorporated modular, which is very good! So let's fix a bit/clarify the model names, fix the consistency issues and then we'll be good for a first review! 🤗

**Comment by FL33TW00D** [comment]
Hey @Cyrilvallez,
Thanks for taking a look! 

Arc has created 2 models, `StateEmbedding` and `StateTransition` which would be good to add. They are the first group to surpass linear baselines for this problem.

We will clean up the PR before pinging for a proper review!

**Comment by Cyrilvallez** [comment]
Nice, thanks for the explanations @FL33TW00D! Makes sense! Let me know when you believe this is ready for review then 🤗 Just a heads-up that we want the folder/files names as full snake_case, e.g. `state_embedding`, and the class name all prefixed with CamelCase, e.g. `StateEmbeddingModule`! 👌

**Comment by abhinadduri** [comment]
thanks everyone! we are starting our review now, cc @Rive-001

File: src/transformers/models/state_tx/configuration_state_tx.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        # Validate the correctness of rotary position embeddings parameters
+        # BC: if there is a 'type' field, copy it it to 'rope_type'.
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+
+
+class StateTxConfig(PretrainedConfig):
+    r"""
+    Configuration class for StateTx (State Transformer) model based on PertSetsPerturbationModel.
```
Comment (Rive-001) [inline src/transformers/models/state_tx/configuration_state_tx.py]: Low priority: We renamed the model class from `PertSetsPerturbationModel` to `StateTransitionPerturbationModel`.

File: src/transformers/models/state_tx/configuration_state_tx.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+    This model uses a bidirectional Llama transformer backbone to process perturbation data.
```
Comment (Rive-001) [inline src/transformers/models/state_tx/configuration_state_tx.py]: Low priority: We currently support bi-directional versions of both Llama and GPT-2.

File: src/transformers/models/state_tx/modeling_state_tx.py [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+from ...modeling_layers import GradientCheckpointingLayer
+from ...modeling_outputs import BaseModelOutputWithPast
+from ...modeling_utils import ALL_ATTENTION_FUNCTIONS, PreTrainedModel
+from ...processing_utils import Unpack
+from ...utils import TransformersKwargs, auto_docstring
+from ...utils.generic import check_model_inputs
+from .configuration_state_tx import LlamaBidirectionalConfig, StateTxConfig
+
+
+class SamplesLoss(nn.Module):
```
Comment (Rive-001) [inline src/transformers/models/state_tx/modeling_state_tx.py]: We currently support MSE and MMD loss functions. We use SampleLoss from the geom library. [SampleLoss documentation](https://www.kernel-operations.io/geomloss/api/pytorch-api.html#geomloss.SamplesLoss)

https://github.com/ArcInstitute/state/blob/be0006c4556327431bda29b6db1b7b223d9eda8c/src/state/tx/models/state_transition.py#L163-L177

https://github.com/ArcInstitute/state/blob/be0006c4556327431bda29b6db1b7b223d9eda8c/src/state/tx/models/state_transition.py#L20-L34

File: src/transformers/models/state_tx/modeling_state_tx.py [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+    """Samples loss function for perturbation model."""
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, predictions, targets):
+        return F.mse_loss(predictions, targets)
+
+
+class LatentToGeneDecoder(nn.Module):
```
Comment (Rive-001) [inline src/transformers/models/state_tx/modeling_state_tx.py]: We currently support user input for number of layers in the decoder, dimensions of those layers and optional residual connections between layers.

https://github.com/ArcInstitute/state/blob/be0006c4556327431bda29b6db1b7b223d9eda8c/src/state/tx/models/base.py#L15-L116

File: src/transformers/models/state_tx/modeling_state_tx.py [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+    """Decoder that converts latent representations back to gene space."""
+
+    def __init__(self, config: StateTxConfig):
+        super().__init__()
+        self.decoder = nn.Sequential(
+            nn.Linear(config.gene_dim, 1024, bias=True),
```
Comment (Rive-001) [inline src/transformers/models/state_tx/modeling_state_tx.py]: This doesn't seem correct, as the input dimension would be a latent dimension and not the gene dimensions. The gene dimension is the output dimension.

https://github.com/ArcInstitute/state/blob/be0006c4556327431bda29b6db1b7b223d9eda8c/src/state/tx/models/base.py#L47

File: src/transformers/models/state_tx/modular_state_tx.py [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+        # Shape: (batch_size, 2, hidden_dim)
+        combined_input = pert_encoded.unsqueeze(1) + basal_encoded.unsqueeze(1)
+        seq_input = combined_input  # (batch_size, 2, hidden_dim)
+
+        # # Add batch embeddings if provided
+        # if batch_ids is not None:
+        #     batch_embeds = self.batch_encoder(batch_ids)  # (batch_size, hidden_dim)
+        #     # Add batch embedding to each position
+        #     combined_input = combined_input + batch_embeds.unsqueeze(1)
+        batch_embeddings = self.batch_encoder(torch.zeros([512]).long()).unsqueeze(1)
```
Comment (Rive-001) [inline src/transformers/models/state_tx/modular_state_tx.py]: Hardcoding batch embeddings to 0s might not be correct.

File: src/transformers/models/state_emb/modeling_state_emb.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+
+        # Gene output decoder
+        self.decoder = nn.Sequential(
+            SkipBlock(config.d_model),
+            nn.Linear(config.d_model, config.output_dim, bias=True),
+        )
+
+        # Binary classification decoder
+        # binary_input_dim = config.output_dim + config.d_model + config.z_dim_rd + config.z_dim_ds
+        binary_input_dim = 4107
```
Comment (Rive-001) [inline src/transformers/models/state_emb/modeling_state_emb.py]: We might want the dimensions of this decoder to be based on the config values.
